Add filters to configure user capability on remote APIs

mehmoodak opened the pull request:
## Description
Added following filters hooks on remote endpoints so that users can configure it as per their needs:

1. `wp_parsely_user_capability_for_all_private_apis` which changes the default user capability for all remote apis
2. `wp_parsely_user_capability_for_[ENDPOINT_FILTER_KEY]_api` which changes the user capability of each API

Closes: #1415 

## How has this been tested?
- Added integration tests.

mehmoodak committed:
➕ ADDS: filter on api to change user capability

mehmoodak committed:
➕ ADDS: integration tests

mehmoodak commented on src/RemoteAPI/class-analytics-post-detail-api.php:
```diff
@@ -18,7 +18,7 @@
  * @since 3.6.0
  */
 class Analytics_Post_Detail_API extends Remote_API_Base {
-	protected const ENDPOINT     = Parsely::PUBLIC_API_BASE_URL . '/analytics/post/detail';
```
- Removed this base URL from here so that we can change the endpoint to filter key in more cleaner way.
- Added this base URL inside `get_api_url` function of Remote_Base_API (only needed once because its a common location)

mehmoodak commented on tests/Integration/Endpoints/AnalyticsPostsProxyEndpointTest.php:
```diff
@@ -140,6 +140,56 @@ public function test_get_items_fails_when_api_secret_is_not_set(): void {
+	public function test_user_is_allowed_to_make_proxy_api_call_if_endpoint_specific_user_capability_is_changed(): void {
+		$this->login_as_contributor();
+		add_filter(
+			'wp_parsely_user_capability_for_analytics_posts_api',
```
Notice that this filter is changed for each remote api.

mehmoodak reviewed (commented):
Added comments for easier review.

mehmoodak committed:
👌 IMPROVE: comments

acicovic commented on src/Utils/utils.php:
```diff
@@ -232,3 +232,16 @@ function convert_to_associative_array( $obj ) {
 }
+
+/**
+ * Convert endpoint to filter key by replacing `/` with `_`.
```
```suggestion
 * Converts endpoint to filter key by replacing `/` with `_`.
```

acicovic commented on tests/Integration/RemoteAPI/RelatedRemoteAPITest.php:
```diff
@@ -69,4 +69,24 @@ public function data_api_url() {
 	}
+
+	/**
+	 * Verifies that public apis does not have filter which checks user capability.
```
```suggestion
	 * Verifies that the endpoint does not have filters that check user capability.
```

acicovic commented on tests/Integration/TestCase.php:
```diff
@@ -445,6 +451,14 @@ public function set_admin_user( $admin_user_id = 1 ): void {
 	}
 
+	/**
+	 * Create a user with role `contributor` and login.
```
```suggestion
	 * Creates a user with role `contributor` and login.
```

acicovic commented on tests/Integration/RemoteAPI/RelatedRemoteAPITest.php:
```diff
@@ -69,4 +69,24 @@ public function data_api_url() {
+	 *
+	 * @uses \Parsely\RemoteAPI\Remote_API_Base::__construct
+	 */
+	public function test_public_endpoint_does_not_have_user_capability_checks(): void {
```
```suggestion
	public function test_related_endpoint_does_not_have_user_capability_filters(): void {
```

acicovic reviewed (commented):
Very nice work!

I think we should have at least one test that checks that when both the default and endpoint-specific filters are set, that the endpoint-specific filter will have the priority.

Or do we have this and I missed it?

mehmoodak committed:
👌 IMPROVE: comments and added test to verify priority of endpoint specific filter

mehmoodak commented:
> I think we should have at least one test that checks that when both the default and endpoint-specific filters are set, that the endpoint-specific filter will have the priority.

Added this test and incorporated other feedback. Please have a look again. Thanks

acicovic commented on tests/Integration/RemoteAPI/AnalyticsPostsRemoteAPITest.php:
```diff
@@ -39,4 +39,75 @@ public function data_api_url() {
+	}
+
+	/**
+	 * Verifies that endpoint specific user capability filter have more priority than default.
```
```suggestion
	 * Verifies that the endpoint specific user capability filter has more priority than the default capability filter.
```

acicovic reviewed (approved):
I think we're good, but left a comment improvement that I think is clearer.

mehmoodak committed:
👌 IMPROVE: comments
